adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -744,41 +978,244 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+                  (f._1.compareTo(partition._1._1) >= 0 && f._1.compareTo(partition._1._2) <= 0)) {
+                  // we use Options here so scala doesn't lose typing information
+                  Some((partition._2, f))
+                } else if (partition._2 == destinationPartitionMap.length - 1 &&
+                  f._1.compareTo(partition._1._2) > 0) {
+                  Some((partition._2, f))
+                  // we do this so we don't filter out any data at the end or beginning
+                  // we don't want to lose data before the join, and this prevents that
+                } else if ((partition._2 == destinationPartitionMap.length - 1 && f._1.compareTo(partition._1._2) > 0) ||
+                  (partition._2 == 0 && f._1.compareTo(partition._1._1) < 0)) {
+                  Some((partition._2, f))
+                } else None
+              })
+            })
+          }, preservesPartitioning = true)
+          // finally we filter out the None values and flatten down
+          // the result is an RDD[(Int, (ReferenceRegion, T))].
+          // the Int is the destination partition index
+          .flatMap(f => f.filter(_.isDefined).map(g => g.get))
+          .map(f => ((f._2._1, f._1), f._2))
+          .repartitionAndSortWithinPartitions(
+            new GenomicPositionRangePartitioner(numPartitions, destinationPartitionMap))
+          .map(_._2)
+      }
+    // here we get the new partition map
+    // for now, we can't use the left RDD's partition map because
+    // the bounds don't always include the values we copartitioned with.
+    // in other words, joins allow data that starts before a given
+    // ReferenceRegion to combine given they overlap.
+    // this means that we cannot use the bounds of the left RDD because
+    // they may not necessarily be the bounds of the right RDD
+    val newPartitionMap = finalPartitionedRDD.mapPartitions(iter => {
+      getPartitionMapBoundsFromRdd(iter)
+    }, preservesPartitioning = true).collect
+
+    replaceRdd(finalPartitionedRDD.values, Some(newPartitionMap))
+  }
+
+  /**
+   * Gets the partition bounds from a ReferenceRegion keyed Iterator
+   *
+   * @param iter The data on a given partition. ReferenceRegion keyed
+   * @return The bounds of the ReferenceRegions on that partition, in an Iterator
+   */
+  private def getPartitionMapBoundsFromRdd(iter: Iterator[(ReferenceRegion, T)]): Iterator[(ReferenceRegion, ReferenceRegion)] = {
+    if (iter.isEmpty) {
+      // this is disappointing
+      // we have to have a null (or empty) value here in the case that our
+      // copartitioned RDD did not have any data that corresponded to this
+      // partition. Spark will squash empty Iterators with a flatten, so
+      // we have to have something here otherwise our partition map will
+      // not have enough partitions.
+      Iterator(null)
+    } else {
+      val head = iter.next._1
```

Comment:
I'd like to avoid the use of `head` and `tail` as variables as they visually shadow methods in the various collection classes.